File: Java/Aplikasi-Search-Github/app/src/main/java/com/makhalibagas/searchgithub/View/Activity/FavoriteActivity.java
package com.makhalibagas.searchgithub.View.Activity;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.annotation.SuppressLint;
import android.database.Cursor;
import android.os.AsyncTask;
import android.os.Bundle;

import com.makhalibagas.searchgithub.Adapter.FavoriteAdapter;
import com.makhalibagas.searchgithub.Database.GithubHelper;
import com.makhalibagas.searchgithub.R;

import java.util.Objects;

import static com.makhalibagas.searchgithub.Database.DatabaseContract.GithubColumns.CONTENT_URI;

public class FavoriteActivity extends AppCompatActivity {

    private Cursor listGithub;
    private FavoriteAdapter favoriteAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favorite);


        GithubHelper githubHelper = GithubHelper.getInstance(getApplicationContext());
        githubHelper.open();


        Objects.requireNonNull(getSupportActionBar()).setTitle(R.string.favorite);
        RecyclerView recyclerView = findViewById(R.id.rvFav);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        favoriteAdapter = new FavoriteAdapter(getApplicationContext());
        favoriteAdapter.setFavoriteList(listGithub);
        recyclerView.setAdapter(favoriteAdapter);
        if (savedInstanceState == null){
            new LoadFavoriteItem().execute();
        }
    }


    @SuppressLint("StaticFieldLeak")
    private class LoadFavoriteItem extends AsyncTask<Void, Void, Cursor>{
        @Override
        protected Cursor doInBackground(Void... voids) {
            return getContentResolver().query(CONTENT_URI, null,null,null,null);
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected void onPostExecute(Cursor cursor) {
            super.onPostExecute(cursor);
            listGithub = cursor;
            favoriteAdapter.setFavoriteList(listGithub);
            favoriteAdapter.notifyDataSetChanged();
        }
    }
}

File: Kotlin/Submission3/favoriteuser/src/main/java/com/makhalibagas/favoriteuser/ui/activity/FavoriteActivity.kt
package com.makhalibagas.favoriteuser.ui.activity

import android.content.Intent
import android.database.ContentObserver
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.makhalibagas.favoriteuser.R
import com.makhalibagas.favoriteuser.adapter.UserAdapter
import com.makhalibagas.favoriteuser.database.UserContract.UserColumns.Companion.CONTENT_URI_USER
import com.makhalibagas.favoriteuser.helper.MappingHelper
import kotlinx.android.synthetic.main.activity_favorite.*
import kotlinx.android.synthetic.main.activity_favorite.btSettings
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

class FavoriteActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_favorite)


        btSettings.setOnClickListener {
            startActivity(Intent(applicationContext, SettingsActivity::class.java))
        }
        val handlerThread = HandlerThread("DataObserver")
        handlerThread.start()
        val handler = Handler(handlerThread.looper)

        val myObserver = object : ContentObserver(handler){
            override fun onChange(selfChange: Boolean) {
                loadUserAsync()
            }
        }

        contentResolver.registerContentObserver(CONTENT_URI_USER, true, myObserver)

        if (savedInstanceState == null){
            loadUserAsync()
        }

        rvFav.layoutManager = LinearLayoutManager(this)
    }

    private fun loadUserAsync(){
        GlobalScope.launch(Dispatchers.Main){
            val deferredUser = async(Dispatchers.IO){
                val cursor = contentResolver.query(CONTENT_URI_USER,null,null,null,null)
                MappingHelper.mapCursorToList(cursor)
            }

            val user = deferredUser.await()
            if (user.isNotEmpty()){
                rvFav.adapter = UserAdapter(applicationContext, user)
            }else{
                notFound.visibility = View.VISIBLE
            }
        }
    }
}
